File: cpp/samchon/protocol/master/ParallelServerClientArrayMediator.cpp
#include <samchon/protocol/master/ParallelServerClientArrayMediator.hpp>

#include <thread>

using namespace std;
using namespace samchon::protocol::external;
using namespace samchon::protocol::master;

/* ---------------------------------------------------------
	CONSTRUCTORS
--------------------------------------------------------- */
ParallelServerClientArrayMediator::ParallelServerClientArrayMediator()
	: ParallelSystemArrayMediator(),
	ExternalServerClientArray()
{
}

ParallelServerClientArrayMediator::~ParallelServerClientArrayMediator()
{
}

/* ---------------------------------------------------------
	OPEN & CONNECT
--------------------------------------------------------- */
void ParallelServerClientArrayMediator::open(int port)
{
	thread t1([this, port]()
	{
		ExternalServerClientArray::open(port);
	});
	thread t2(&ParallelServerClientArrayMediator::start_mediator, this);

	t1.join();
	t2.join();
}

void ParallelServerClientArrayMediator::connect()
{
	thread t1([this]()
	{
		ExternalServerClientArray::connect();
	});
	thread t2(&ParallelServerClientArrayMediator::start_mediator, this);

	t1.join();
	t2.join();
}

File: ts/src/samchon/protocol/master/ParallelServerClientArrayMediator.ts
/// <reference path="../../API.ts" />

/// <reference path="ParallelClientArrayMediator.ts" />

namespace samchon.protocol.master
{
	export abstract class ParallelServerClientArrayMediator
		extends ParallelClientArrayMediator
		implements external.IExternalServerClientArray
	{
		/* ---------------------------------------------------------
			CONSTRUCTORS
		--------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();
		}

		protected createChild(xml: library.XML): ParallelSystem
		{
			return this.createExternalServer(xml);
		}
		protected abstract createExternalServer(xml: library.XML): IParallelServer;

		/* ---------------------------------------------------------
			METHOD OF CLIENT
		--------------------------------------------------------- */
		public connect(): void
		{
			for (let i: number = 0; i < this.size(); i++)
				if (this.at(i)["connect"] != undefined)
					(this.at(i) as external.ExternalServer).connect();
		}
	}
}